### DISCUSSION TITLE
expose_function function

### DISCUSSION THREAD

**Discussion by 3mora2** [seleniumbase/SeleniumBase#1924] [General]
Hello, there is way to add function like `expose_function`, it in [playwright](https://playwright.dev/python/docs/api/class-page#page-expose-function) ?

> The method adds a function called `name` on the `window` object of every frame in the page. When called, the
        function executes `callback` and returns a [Promise] which resolves to the return value of `callback`.

**Comment by mdmintz** [comment]
You can achieve the same effect by using ``self.execute_script()`` in seleniumbase:

```python
import hashlib
from seleniumbase import BaseCase
BaseCase.main(__name__, __file__)

class MyTest(BaseCase):
    def sha256(self, text):
        m = hashlib.sha256()
        m.update(bytes(text, "utf8"))
        return m.hexdigest()

    def test_expose(self):
        code = self.sha256("hello")
        self.execute_script("window.code='%s';" % code)
        print(self.execute_script("return window.code;"))
```

If you just need to pass variables around, you can use standard Python for that. You can also save variables to `self`.

**Reply by 3mora2** [reply]
Thanks for the response, `expose_function ` used for callback, i mean set js callback for event like receive new message then call callback function to do things.
Scenario like:
> set callback python function `receive_message` into js
when new message receive in js 
> > js return call python function `receive_message`
print new message in python

**Reply by mdmintz** [reply]
It can all be done via ``self.execute_script()``. I would need to see a very specific example to show you a SeleniumBase equivalent.

**Reply by 3mora2** [reply]
```
from playwright.sync_api import sync_playwright


def receive_message(*args, **kwargs):
    """callback function"""
    print(args, kwargs)


# Open page
playwright = sync_playwright().start()
browser = playwright.chromium.launch(channel="chrome", headless=False)
context = browser.new_context(no_viewport=True)
page = context.new_page()

page.goto("https://www.google.com/")
# set callback function
page.expose_function("receive_message", receive_message)
# add for listen for events
page.wait_for_timeout(1000000)
```
![Screenshot_1](https://github.com/seleniumbase/SeleniumBase/assets/66757189/a3771e39-cad6-4bef-a67e-489c29cf89cc)

**Reply by mdmintz** [reply]
The alternative: https://github.com/seleniumbase/SeleniumBase/blob/master/examples/test_console_logging.py

Related methods:

```python
self.start_recording_console_logs()
self.console_log_string(string)
self.console_log_script(script)
self.get_recorded_console_logs()
```

Create a loop, read console logs, and perform an action in Python if a string is found.

**Reply by 3mora2** [reply]
calling function from console,it is as example,
I try convert [wppconnect](https://github.com/wppconnect-team/wppconnect) to python [WPP_Whatsapp](https://github.com/3mora2/WPP_Whatsapp), by using playwright, but it is not thread safe, so I am try [SeleniumBase](https://github.com/seleniumbase/SeleniumBase/tree/master) and need:

- listen for driver reload page to inject js to whatsapp [here](https://github.com/3mora2/WPP_Whatsapp/blob/main/Whatsapp/api/layers/HostLayer.py#L70)
- listen for event like (new mesage, status,....) receive from js code 
  - create `handel_func` in python for listen js event and handel receive data [](https://github.com/3mora2/WPP_Whatsapp/blob/main/Whatsapp/api/layers/ListenerLayer.py#L74)
  -  expose python function `handel_func` to js window as `window['onMessage']` for example [here](https://github.com/3mora2/WPP_Whatsapp/blob/main/Whatsapp/api/layers/ListenerLayer.py#L75)
  - then add js listen for new message, if receive message call `window['onMessage'](object)` refere to `handel_func` [here](https://github.com/3mora2/WPP_Whatsapp/blob/main/Whatsapp/api/layers/ListenerLayer.py#L86)
